internal/pkg/earpointer/taint.go

```diff
@@ -206,33 +211,52 @@ type SourceSinkTrace struct {
 
 // Look for <source, sink> pairs by examining the heap alias information.
 func SourcesToSinks(funcSources source.ResultType, isTaintField func(named *types.Named, index int) bool,
-	heap *Partitions, conf *config.Config) []*SourceSinkTrace {
+	heap *Partitions, conf *config.Config) map[ssa.Instruction]*SourceSinkTrace {
 
-	var traces []*SourceSinkTrace
+	calleeMap := mapCallees(heap.cg) // map a callsite to the possible callees
+	traces := make(map[ssa.Instruction]*SourceSinkTrace)
 	for fn, sources := range funcSources {
-		// Transitively get the set of functions called within "fn".
+		// Transitively get the set of functions reachable from "fn".
 		// This set is used to narrow down the set of references needed to be
 		// considered during EAR heap traversal. It can also help reducing the
 		// false positives and boosting the performance.
-		callees := boundedDepthCallees(fn, conf.EARTaintCallSpan)
+		// For example,
+		//   func f1(){ f2(); g4() }
+		//   func f2(){ g1(); g2() }
+		//   func g1(){ g3() }
+		// g1 can reach f2 through the caller, and then f1 similarly,
+		// and then g4 through the callee.
+		// g1's full reachable set is {f1,f2,g1,g2,g3,g4}.
+		reachable := boundedReachableFunctions(fn, heap.cg, conf.EARTaintCallSpan)
+		// Start from the set of taint sources.
 		srcRefs := make(map[*source.Source]ReferenceSet)
 		for _, s := range sources {
-			srcRefs[s] = srcAliasRefs(s, isTaintField, heap, callees)
+			srcRefs[s] = srcAliasRefs(s, isTaintField, heap, reachable)
 		}
-		// Traverse all the callee functions (not just the ones with sink sources)
-		ht := &heapTraversal{heap: heap, callees: callees, visited: make(ReferenceSet)}
-		for member := range callees {
+		// Traverse all the reachable functions (not just the ones with sink sources)
+		// in search for connected sinks.
+		ht := &heapTraversal{heap: heap, callees: reachable, visited: make(ReferenceSet)}
+		for member := range reachable {
 			for _, b := range member.Blocks {
 				for _, instr := range b.Instrs {
 					switch v := instr.(type) {
 					case *ssa.Call:
+						callees := calleeMap[&v.Call]
+						if callees == nil { // the static call graph may ignore some cases
+							if sc := v.Call.StaticCallee(); sc != nil {
```

Comment:
This seems unlikely to help. There is a static callee, but no edge in the static call graph with depth >= 1.

Can you point to a case it does help? If it does, this seems like a bug and probably deserves a todo.

Reply:
Without this one unit test fails:

../levee/testdata/src/levee_analysistest/example/tests/receivers/tests.go

```
type sourceBuilder struct {
  ...
}

func (b *sourceBuilder) buildP() {
  ...
}
```
Method `buildP` is not in the call graph. It is because it is not included in the method set used by the following code: 

https://github.com/golang/tools/blob/master/go/ssa/ssautil/visit.go#L44-L45

This seems to be a bug in SSA. A temporary work-around is to expose "sourceBuilder" by renaming it to "SourceBuilder" as done here.

Reply:
Missing a function is concerning. If we are not under time pressure, I would like to understand the problem better before proceeding instead of hacking around it.

How does the unit test fail? Which package is being analyzed? Why does exposing the function fix this?

Reply:
An internal bug report with a reproducer has been created in Buganizer.

It seems that the ssa builder may ignore some unexposed methods:
 
https://github.com/golang/tools/blob/master/go/ssa/builder.go#L2278-L2279

Reply:
ssautil was producing a list of functions that correctly concluded that sourceBuilder and these methods were unreachable. (There are other concerns for other cases, but this case was a correct conclusion.) I think we want this code to rely on the callgraph it is given, and the most recent version does this.

Changing to "SourceBuilder" is a good fix in this case. I think this has been sufficiently addressed.